### PR TITLE
Use designspace source names (replacement PR for #367)

### PR DESCRIPTION
**Pull request by verbosus**
Here’s a first stab at an updated PR from @moyogo’s #367. I’ve added a bunch of tests to excercise all the various code paths.

### CODE REVIEW

**Comment by verbosus** [comment]
Oooops, need to lint the code with `black`, please hold.

**Comment by verbosus** [comment]
I don’t understand the lint error message. I ran `black` on the sources a few times but it still fails.

**Comment by anthrotype** [comment]
Maybe the version on black which is running on the CI is different from the one you have installed locally

**Comment by anthrotype** [comment]
It's isort that is failing apparently, not black. You can run all the lint checks locally using `tox -e lint`. You can edit the sources manually to accommodate the linter. Not sure if isort can automatically reformat sources

**Comment by verbosus** [comment]
Thanks, will do.

**Comment by verbosus** [comment]
I’m on Python 3.9.5, here are my Python packages including `isort` which is the same version as the server, 5.9.3:

```
appdirs==1.4.4
attrs==21.2.0
backports.entry-points-selectable==1.1.0
black==21.7b0
booleanOperations==0.9.0
cffsubr==0.2.8
click==8.0.1
compreffor==0.5.1
coverage==5.5
cu2qu==1.6.7
defcon==0.8.1
distlib==0.3.2
filelock==3.0.12
flake8==3.9.2
flake8-bugbear==21.4.3
fonttools==4.26.1
fs==2.4.13
iniconfig==1.1.1
isort==5.9.3
lxml==4.6.3
mccabe==0.6.1
mypy-extensions==0.4.3
packaging==21.0
pathspec==0.9.0
platformdirs==2.2.0
pluggy==0.13.1
py==1.10.0
pyclipper==1.3.0
pycodestyle==2.7.0
pyflakes==2.3.1
pyparsing==2.4.7
pytest==6.2.4
pytz==2021.1
regex==2021.8.21
six==1.16.0
skia-pathops==0.6.0.post2
toml==0.10.2
tomli==1.2.1
tox==3.24.3
-e git+ssh://git@github.com/verbosus/ufo2ft.git@52c3c0d456f86558fe6d3f42c6776b69f5f16433#egg=ufo2ft
ufoLib2==0.7.1
virtualenv==20.7.2
```

Here’s the isort output on my Mac if ran on the root directory:

```
$ isort --check-only --diff .
Skipped 3 files
```

If I run the command from inside the `Lib/ufo2ft` directory I get no output.

Can you repro the issue @anthrotype @moyogo?

**Comment by verbosus** [comment]
Here’s something funny: if I run `tox -e lint` on my branch everything is fine and the test passes. If I run it on `main`, the test fails. 

I looked into what the server is actually doing, and it seems to be doing the right thing: it’s checking out main, merging my branch in and running the tests on _that._ I reproduced the exact commands from the Test + Deploy / lint (pull_request) action:

```
(ufo2ft):~/S/ufo2ft-orig:main$ 
git -c protocol.version=2 fetch --no-tags --prune --progress --no-recurse-submodules --depth=1 origin +591693e2fe7d1fef4dcbbedb7d28685d2f7c5a36:refs/remotes/pull/527/merge
remote: Total 0 (delta 0), reused 0 (delta 0), pack-reused 0
(ufo2ft):~/S/ufo2ft-orig:main$ git checkout --progress --force refs/remotes/pull/527/merge
Note: switching to 'refs/remotes/pull/527/merge'.

You are in 'detached HEAD' state. You can look around, make experimental
changes and commit them, and you can discard any commits you make in this
state without impacting any branches by switching back to a branch.

If you want to create a new branch to retain commits you create, you may
do so (now or later) by using -c with the switch command. Example:

  git switch -c <new-branch-name>

Or undo this operation with:

  git switch -

Turn off this advice by setting config variable advice.detachedHead to false

HEAD is now at 591693e Merge 52c3c0d456f86558fe6d3f42c6776b69f5f16433 into 0e8edb5a4c20648a52f314103e2c1f0cdb0d2cf9
(ufo2ft):~/S/ufo2ft-orig:(HEAD detached at pull/527/merge)$ tox -e lint
lint installed: appdirs==1.4.4,attrs==21.2.0,black==21.7b0,click==8.0.1,coverage==5.5,flake8==3.9.2,flake8-bugbear==21.4.3,iniconfig==1.1.1,isort==5.9.3,mccabe==0.6.1,mypy-extensions==0.4.3,packaging==21.0,pathspec==0.9.0,pluggy==0.13.1,py==1.10.0,pycodestyle==2.7.0,pyflakes==2.3.1,pyparsing==2.4.7,pytest==6.2.4,regex==2021.8.21,toml==0.10.2,tomli==1.2.1
lint run-test-pre: PYTHONHASHSEED='973146273'
lint run-test: commands[0] | black --check --diff .
All done! ✨ 🍰 ✨
52 files would be left unchanged.
lint run-test: commands[1] | isort --check-only --diff .
Skipped 2 files
lint run-test: commands[2] | flake8
________________________________________________________________ summary _________________________________________________________________
  lint: commands succeeded
  congratulations :)
```

**Comment by verbosus** [comment]
Still don’t quite understand what’s happening, in the meantime I found this: [inconsistent isorting sorting between local and CI execution #1790](https://github.com/PyCQA/isort/issues/1790) which seems not far from what I’m experiencing.

**Comment by verbosus** [comment]
I figured it out. The root directory _must_ be called `ufo2ft` for the tests to pass. I often run multiple parallel clones of the same repo side-by-side for easy comparisons, etc. and was working from a directory called `ufo2ft-debug`. I found this whole process rather confusing. Admittedly, it’s the first time I’ve ever been exposed to `isort` and so it’s all very likely due to my inexperience, but I wouldn’t have thought my root directory name to matter. It shouldn’t, especially because our code lives inside `Lib/ufo2ft`.

**Comment by verbosus** [comment]
I think this PR is now ready for review. Let me know @anthrotype @moyogo

**Comment by anthrotype** [comment]
> The root directory must be called ufo2ft for the tests to pass

I didn't know that either (@madig hooked up isort to ufo2ft linters). Yeah, that's a bit weird but I guess we can live with it.

I'll review the rest of the PR now

**Review by anthrotype — COMMENTED** [review]
Thanks Antonio for picking this up!

I left a couple of comments.

One thing about the unit tests. I would prefer if we make the test more tailored to the specific functionality they are testing and that we don't add many UFOs to the test data -- unless when really required (e.g. when we do integration testing and we want to run ufo2ft end-to-end).
Adding a bunch of whole fonts (with lots of lines that nobody actually ever reads) just for testing a very specific piece of code it kind of hides things and makes it more cumbersome for the code and test suite to evolve. 
In this case, you're interested in testing this specific functionality of overriding family/style name from designspace, so you don't actually need to add entire fonts just for that.
You could create minimal font objects from code as test fixture without checking them in as files under VCS.
(I imagine it took you some time to actually make the test fonts, so apologies for that)

File: Lib/ufo2ft/util.py [hunk 1]
```diff
@@ -4,6 +4,7 @@
 from copy import deepcopy
 from inspect import getfullargspec
 
+import defcon
```
[the remark this answers is not in the document]
Comment (verbosus) [reply Lib/ufo2ft/util.py]: Fixed thanks.

File: Lib/ufo2ft/__init__.py [hunk 2]
```diff
@@ -415,7 +419,14 @@ def compileInterpolatableTTFsFromDS(
                 "designspace source '%s' is missing required 'font' attribute"
                 % getattr(source, "name", "<Unknown>")
             )
-        ufos.append(source.font)
+
+        (
+            _source_font,
+            _family_modified,
+            _style_modified,
```
[the remark this answers is not in the document]
Comment (verbosus) [reply Lib/ufo2ft/__init__.py]: Fixed thanks.

File: Lib/ufo2ft/util.py [hunk 3]
```diff
@@ -484,6 +485,36 @@ def _loadPluginFromString(spec, moduleName, isValidFunc):
     return klass(**options)
 
 
+def overrideSourceNamesFromDesignSpace(font, source, inplace=False):
+    """Allow overriding the family and style name in a source font
+    if its designspace entry specifies custom values"""
+    _source_font = font
```
[the remark this answers is not in the document]
Comment (verbosus) [reply Lib/ufo2ft/util.py]: Okay.

**Comment by verbosus** [comment]
@anthrotype I addressed a few of your comments. I need to look into refactoring the tests without using the UFOs I had already made.

File: Lib/ufo2ft/util.py [hunk 4]
```diff
@@ -485,34 +484,31 @@ def _loadPluginFromString(spec, moduleName, isValidFunc):
     return klass(**options)
 
 
-def overrideSourceNamesFromDesignSpace(font, source, inplace=False):
-    """Allow overriding the family and style name in a source font
+def overrideSourceNamesFromDesignSpace(font, designspace_entry, inplace=False):
+    """Allow overriding the family and style name in a font
     if its designspace entry specifies custom values"""
-    _source_font = font
+    font_class = font.__class__
```
Comment (anthrotype) [inline Lib/ufo2ft/util.py:490]: nit: you only use `font.__class__` inside the defcon-related if block, you could just move this line in there

Comment (anthrotype) [reply Lib/ufo2ft/util.py:490]: I mean, just under `if not inplace:`

Comment (verbosus) [reply Lib/ufo2ft/util.py]: Fixed thanks.

**Comment by verbosus** [comment]
Converting PR back to draft as I refactor the tests.

**Comment by verbosus** [comment]
Back to ready for review. Can you take a look @anthrotype? Thanks.

File: Lib/ufo2ft/util.py [hunk 5]
```diff
@@ -484,6 +484,33 @@ def _loadPluginFromString(spec, moduleName, isValidFunc):
...
+
+    if not inplace:
+        font_class = font.__class__
+        if hasattr(font_class, "getDataForSerialization"):
+            # work around defcon.Font deepcopy
+            data = font.getDataForSerialization()
+            font = font_class()
+            font.setDataFromSerialization(data)
+        else:
+            font = deepcopy(font)
```
Comment (anthrotype) [inline Lib/ufo2ft/util.py:499]: you're doing the copy (which is relatively expensive operation) before knowing if the copy is needed, that is before checking if the designspace overrides familyName or styleName.
I suggest you first check that and only copy if either is not None (and inplace is False of course)

File: tests/variable_test.py [hunk 6]
```diff
@@ -0,0 +1,353 @@
...
+        s1.styleName = source_style_name_override
+    doc.addSource(s1)
+
+    s2 = SourceDescriptor()
+    s2.name = "master.TestFontNames.Heavy.800"
+    s2.font = heavy_font()
+    s2.location = dict(weight=800)
+    doc.addSource(s2)
+
+    if instances:
```
Comment (anthrotype) [inline tests/variable_test.py:98]: the `if instances:` is superfluous, for loop will just skip if `instances` tuple is empty

File: tests/variable_test.py [hunk 7]
```diff
@@ -0,0 +1,353 @@
...
+        for i in instances:
+            doc.addInstance(i)
+
+    return doc
+
+
+def make_designspace_plain(
```
Comment (anthrotype) [inline tests/variable_test.py:105]: what's the use of `make_designspace_plain` function? it does exactly the same thing as `make_designspace` function, you can simply remove the former and replace it with calls to the latter

File: tests/variable_test.py [hunk 8]
```diff
@@ -0,0 +1,353 @@
+import defcon
```
Comment (anthrotype) [inline tests/variable_test.py:1]: you need not import defcon directly here. You can use the `ufo_module` global fixture defined in tests/conftest.py and already available to all pytest tests under tests/ folder. That way tests will be run twices, once with defcon, the other time with ufoLib2 (so you test both).
Actually you only import defcon module to initialise a Font object, so you want to use the `FontClass` fixture also defined inside tests/conftest.py module.
Note that you don't need to explicitly import the fixture in conftest.py, that's a special module that pytest looks for and automatically makes all fixture defined in it available for tests defined in the same folder tree.
grep for `FontClass` in the tests/ folder to see examples of how it's used.

**Review by anthrotype — COMMENTED** [review]
thanks for working on refactoring the tests. I left a few other comments below.

I was actually thinking tht we just add a unit test for `overrideSourceNamesFromDesignSpace` inside a tests/util_test.py that only tests that function. We don't actually need to tests the whole pipeline, that's tested elsewhere (in integration_test.py)

File: tests/variable_test.py (hunk 7 above)
Comment (anthrotype) [reply tests/variable_test.py:105]: ah, I see, it populates `instances`.

File: tests/variable_test.py (hunk 6 above)
Comment (verbosus) [reply tests/variable_test.py:98]: Fixed.

File: Lib/ufo2ft/util.py (hunk 5 above)
Comment (verbosus) [reply Lib/ufo2ft/util.py:499]: Fixed.

File: tests/variable_test.py (hunk 8 above)
Comment (verbosus) [reply tests/variable_test.py]: I fixed this in b0e26d67ded01eccb22ea87eb53e459475b964b6

**Comment by verbosus** [comment]
Looks like some of the designspace v5 work has addressed this elsewhere, because it’s now working correctly. So no need to keep the PR up anymore. Brilliant!

**Comment by verbosus** [comment]
Aaaaand we’re back. The family / style name overriding still does not work in case of a variable font, and it’s dubious whether it will in case of a static instance. I got confused because it works in `fontmake` thanks to the overriding taking places in its instancer, but it might still be broken here. So we need to revise the tests and still send this PR through, I think. I’ll do that and report back.

**Pull request closed without merging**